gtsfm/utils/metrics.py

```diff
@@ -55,21 +62,152 @@ def count_correct_correspondences(
+    back-projected at each keypoint in the images and intersections between these rays and the ground truth mesh are
+    recorded. Next, given a match, the the mesh intersections corresponding to each keypoint is forward-projected into
+    the other image and the reprojection error is computed to decide whether the match is an inlier.
+
+    Args:
+        keypoints_i1: N keypoints in image i1.
+        keypoints_i2: N corresponding keypoints in image i2.
+        gt_camera_i1: ground truth camera for image i1, i.e., wTi1 and intrinsics.
+        gt_camera_i1: ground truth camera for image i2, i.e., wTi2 and intrinsics.
+        gt_scene_mesh: ground truth triangular surface mesh of the scene in the world frame.
+        dist_threshold: max acceptable distance (in pixels) between image coordinates of ground truth landmark and
+            keypoint.
+
+    Returns:
+        is_inlier: (N, ) mask of inlier correspondences.
+        reproj_err: maximum error between forward-projected ground truth landmark and corresponding keypoints
+
+    Raises:
+        ValueError if the number of keypoints do not match.
+    """
+    if len(keypoints_i1) != len(keypoints_i2):
+        raise ValueError("Keypoints must have same counts")
+    n_corrs = len(keypoints_i1)
+    is_inlier = np.zeros(n_corrs, dtype=bool)
+
+    # Perform ray tracing to compute keypoint intersections.
+    keypoint_ind_i1, intersections_i1 = compute_keypoint_intersections(keypoints_i1, gt_camera_i1, gt_scene_mesh)
+    keypoint_ind_i2, intersections_i2 = compute_keypoint_intersections(keypoints_i2, gt_camera_i2, gt_scene_mesh)
+    keypoint_ind, i1_idx, i2_idx = np.intersect1d(keypoint_ind_i1, keypoint_ind_i2, return_indices=True)
+
+    # Forward project intersections into other image to compute error.
+    reproj_err = np.array([np.nan] * len(keypoints_i1))
+    for i in range(len(keypoint_ind)):
+        uv_i1 = keypoints_i1.coordinates[keypoint_ind[i]]
+        uv_i2 = keypoints_i2.coordinates[keypoint_ind[i]]
+        uv_i2i1, success_flag_i1 = gt_camera_i1.projectSafe(intersections_i2[i2_idx[i]])
+        uv_i1i2, success_flag_i2 = gt_camera_i2.projectSafe(intersections_i1[i1_idx[i]])
+        if success_flag_i1 and success_flag_i2:
+            err_i2i1 = np.linalg.norm(uv_i1.flatten() - uv_i2i1.flatten())
+            err_i1i2 = np.linalg.norm(uv_i2.flatten() - uv_i1i2.flatten())
+            is_inlier[keypoint_ind[i]] = max(err_i2i1, err_i1i2) < dist_threshold
+            reproj_err[keypoint_ind[i]] = max(err_i2i1, err_i1i2)
+        else:
+            is_inlier[keypoint_ind[i]] = False
+            reproj_err[keypoint_ind[i]] = np.nan
+
+    return is_inlier, reproj_err
+
+
+def compute_keypoint_intersections(
+    keypoints: Keypoints, gt_camera: PinholeCameraCal3Bundler, gt_scene_mesh: trimesh.Trimesh, verbose: bool = False
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Computes intersections between ground truth surface mesh and rays originating from image keypoints.
```

Comment:
do we not need Embree anymore? or does `trimesh` use it by default, so it doesn't need to be installed here?

It would be nice in this docstring to explain what library is performing the raycasting routine.

Reply:
It can be run with or without Embree. However, it's orders of magnitude faster with it.

Reply:
I see. Are we running it with Embree enabled? It would be nice to include this info in the docstring.

Reply:
This is defined when the mesh is read in with the `use_embree` option. So, this function just goes with whatever the loader defines.

For the `AstronetLoader`, Embree is used.